Divide by zero when scrape_config is not specified

When running the agent with a bare config that includes a `scrape_config` but that omits specifying `scrape_interval`, I see a crash due to the zero scrap interval:

```
panic: runtime error: integer divide by zero

goroutine 291 [running]:
github.com/prometheus/prometheus/scrape.(*Target).offset(0xc000409900, 0x0, 0xeab27cf2a6289c0a, 0x0)
	/home/rob/grafana/agent/vendor/github.com/prometheus/prometheus/scrape/target.go:161 +0x12d
github.com/prometheus/prometheus/scrape.(*scrapeLoop).run(0xc000409a40, 0x0, 0x0, 0x0)
	/home/rob/grafana/agent/vendor/github.com/prometheus/prometheus/scrape/scrape.go:914 +0x77
created by github.com/prometheus/prometheus/scrape.(*scrapePool).sync
	/home/rob/grafana/agent/vendor/github.com/prometheus/prometheus/scrape/scrape.go:423 +0x6ef
```

For reference, the config is

```
server:
  http_listen_address: localhost
  http_listen_port: 9898

prometheus:
  wal_directory: ./wal
  configs:
    - name: myconfig
      remote_write:
        - url: https://prometheus-us-central1.grafana.net/api/prom/push
          basic_auth:
            username: xxx
            password: yyy
      scrape_configs:
        - job_name: 'node'
          static_configs:
            - targets: ['localhost:9100']
```

(I'll submit a PR with a tentative fix; thought I'd file this separately anyway for documentation purposes.)

--- a/pkg/prom/agent.go
+++ b/pkg/prom/agent.go
@@ -62,11 +62,10 @@ func (c *Config) ApplyDefaults() error {
 
 	usedNames := map[string]struct{}{}
 
-	for i, cfg := range c.Configs {
-		err := c.Configs[i].ApplyDefaults(&c.Global)
-		if err != nil {
+	for i := range c.Configs {
+		name := c.Configs[i].Name
+		if err := c.Configs[i].ApplyDefaults(&c.Global); err != nil {
 			// Try to show a helpful name in the error
-			name := cfg.Name
 			if name == "" {
 				name = fmt.Sprintf("at index %d", i)
 			}
@@ -74,13 +73,13 @@ func (c *Config) ApplyDefaults() error {
 			return fmt.Errorf("error validating instance %s: %w", name, err)
 		}
 
-		if _, ok := usedNames[cfg.Name]; ok {
+		if _, ok := usedNames[name]; ok {
 			return fmt.Errorf(
 				"prometheus instance names must be unique. found multiple instances with name %s",
-				cfg.Name,
+				name,
 			)
 		}
-		usedNames[cfg.Name] = struct{}{}
+		usedNames[name] = struct{}{}
 	}
 
 	for i := range c.Configs {

--- a/pkg/prom/agent.go
+++ b/pkg/prom/agent.go
@@ -35,6 +35,7 @@ var (
 
 var (
 	DefaultConfig = Config{
+		Global:                 config.DefaultGlobalConfig,
 		InstanceRestartBackoff: 5 * time.Second,
 	}
 )
@@ -48,6 +49,13 @@ type Config struct {
 	ServiceClientConfig    client.Config       `yaml:"scraping_service_client"`
 	Configs                []instance.Config   `yaml:"configs,omitempty"`
 	InstanceRestartBackoff time.Duration       `yaml:"instance_restart_backoff,omitempty"`
+}
+
+func (c *Config) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultConfig
+
+	type plain Config
+	return unmarshal((*plain)(c))
 }
 
 // ApplyDefaults applies default values to the Config and validates it.

--- a/pkg/prom/agent.go
+++ b/pkg/prom/agent.go
@@ -81,14 +81,6 @@ func (c *Config) ApplyDefaults() error {
 		}
 		usedNames[name] = struct{}{}
 	}
-
-	for i := range c.Configs {
-		err := c.Configs[i].ApplyDefaults(&c.Global)
-		if err != nil {
-			return err
-		}
-	}
-
 	return nil
 }
 

--- a/pkg/prom/agent_test.go
+++ b/pkg/prom/agent_test.go
@@ -82,6 +82,31 @@ func copyConfig(t *testing.T, c Config) Config {
 	return cp
 }
 
+func TestConfigNonzeroDefaultScrapeInterval(t *testing.T) {
+	cfgText := `
+wal_directory: ./wal
+configs:
+  - name: testconfig
+    scrape_configs:
+      - job_name: 'node'
+        static_configs:
+          - targets: ['localhost:9100']
+`
+
+	var cfg Config
+
+	err := yaml.Unmarshal([]byte(cfgText), &cfg)
+	require.NoError(t, err)
+	err = cfg.ApplyDefaults()
+	require.NoError(t, err)
+
+	require.Equal(t, len(cfg.Configs), 1)
+	instanceConfig := cfg.Configs[0]
+	require.Equal(t, len(instanceConfig.ScrapeConfigs), 1)
+	scrapeConfig := instanceConfig.ScrapeConfigs[0]
+	require.Greater(t, int64(scrapeConfig.ScrapeInterval), int64(0))
+}
+
 func TestAgent(t *testing.T) {
 	// Lanch two instances
 	cfg := Config{

--- a/pkg/prom/instance/instance.go
+++ b/pkg/prom/instance/instance.go
@@ -79,11 +79,7 @@ func (c *Config) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	*c = DefaultConfig
 
 	type plain Config
-	if err := unmarshal((*plain)(c)); err != nil {
-		return err
-	}
-
-	return nil
+	return unmarshal((*plain)(c))
 }
 
 func (c Config) MarshalYAML() (interface{}, error) {

--- a/pkg/prom/wal/wal.go
+++ b/pkg/prom/wal/wal.go
@@ -149,7 +149,7 @@ func NewStorage(logger log.Logger, registerer prometheus.Registerer, path string
 }
 
 func (w *Storage) replayWAL() error {
-	level.Info(w.logger).Log("msg", "replaying WAL, this may take awhile", "dir", w.wal.Dir())
+	level.Info(w.logger).Log("msg", "replaying WAL, this may take a while", "dir", w.wal.Dir())
 	dir, startFrom, err := wal.LastCheckpoint(w.wal.Dir())
 	if err != nil && err != record.ErrNotFound {
 		return errors.Wrap(err, "find last checkpoint")

